mipengine/controller/controller.py

```diff
@@ -57,11 +57,12 @@ async def exec_algorithm(
     ):
         context_id = get_a_uniqueid()
         algo_execution_logger = ctrl_logger.get_request_logger(request_id=request_id)
-        algo_execution_logger.info(
-            f"Experiment with request id '{request_id}' started, "
-            f"with algorithm '{algorithm_name}', "
-            f"touching datasets '{','.join(algorithm_request_dto.inputdata.datasets)}', "
-            f"with parameters '{algorithm_request_dto.json()}'."
+        _log_experiment_execution(
+            logger=algo_execution_logger,
+            request_id=request_id,
+            algorithm_name=algorithm_name,
+            datasets=algorithm_request_dto.inputdata.datasets,
+            algorithm_parameters=algorithm_request_dto.json(),
```

Comment:
1. The method is called `exec_algorithm` so I though it's safe to assume it starts here. I think preparing nodes/task_handlers is part of the execution from a third party's perspective. Do you want me to add the log somewhere else?
2. There is no clean way of making these strings dependent on one another. The tests will cover the case when something changes.

Reply:
Forgot to mention on (2) that this script needs to be able to run standalone. As you can see it's added in the mipdb container as well. In the future it could be integrated in MIP or other places, so it should have no dependencies on MIP-Engine.

Reply:
@apmariglis Extracted the federation info logs to separate module as discussed.

Reply:
Creating a separate `mipengine/controller/federation_info_logs.py` was not what I meant. What I meant was, that since you already have a module called `controller_logger.py`, all functionality relevant to "controller logging" should go there. I would completely remove `controller_logger.py` and have a new module `logging.py` like that
```
class Logger:

      def log(self,text:str):
          ...

      def debug(self,text:str):
          ...

      def warning(self,text:str):
          ...

class ExperimentLogger(Logger):

      def __init__(self,request_id):
          ...

      def log_experiment_start(self,algorithm_name,datasets,algorithm_parameters):
          ...

class NodeRegistryLogger(Logger):

      def __init__(self):
          ...

      def log_node_left_federation(node):
          ...

      def log_node_joined_federation(node):
          ...

      def log_datamodel_removed(data_model):
          ...

      def log_datamodel_added(data_model):
          ...

      def log_dataset_added(data_model, dataset,  new_datasets_per_data_model):
          ...

      def log_dataset_removed(data_model, dataset, old_datasets_per_data_model):
          ...
```
Also, the existing function `controller_logger.py::get_background_service_logger` does not make much sense, because node registry (as well as the node_landscape_aggregator) is not a service, as it is not implemented as a service. What you re merely trying to achieve there is to write to the same same log stream (be it a file, or whatever) that the other components write. To achieve that you could implement a factory class that provides the relevant loggers instead of instantiating them in the Controller, NodeRegistry or wherever a logger is needed.
```
class LoggerFactory(metaclass=Singleton):

      def get_experiment_logger(request_id):
          ...
          return ExperimentLogger(...)

      def get_node_registry_logger():
          ...
          return NodeRegistryLogger(...)

      def get_generic_logger(...):
          ...
          return Logger(...)
```         
So then, anywhere we need a logger we get the appropriate logger from the LoggerFactory.